test/test_torch.py

```diff
@@ -12561,61 +12579,70 @@ def binary_check_mem_overlap(self, inplace_op, value=-0.5, device='cpu'):
-        TestTorch.unary_check_mem_overlap(self, lambda t: t.log2_(), device=device)
-        TestTorch.unary_check_mem_overlap(self, lambda t: t.round_(), device=device)
-        TestTorch.unary_check_mem_overlap(self, lambda t: t.rsqrt_(), device=device)
-        TestTorch.unary_check_mem_overlap(self, lambda t: t.sin_(), device=device)
-        TestTorch.unary_check_mem_overlap(self, lambda t: t.sqrt_(), device=device)
-        TestTorch.unary_check_mem_overlap(self, lambda t: t.tan_(), device=device)
-        TestTorch.unary_check_mem_overlap(self, lambda t: t.tanh_(), device=device)
-        TestTorch.unary_check_mem_overlap(self, lambda t: t.trunc_(), device=device)
-
-    @staticmethod
-    def _test_inplace_binary_mem_overlap(self, device='cpu'):
-        binary_ops = ['add_', 'mul_', 'div_', 'sub_']
-        for op in binary_ops:
-            TestTorch.binary_check_mem_overlap(self, op, device=device)
-
-    def test_inplace_unary_mem_overlap(self):
-        return self._test_inplace_unary_mem_overlap(self)
-
-    def test_inplace_binary_mem_overlap(self):
-        return self._test_inplace_binary_mem_overlap(self)
+    def binary_check_input_output_mem_overlap(self, op, device):
+        sz = 3
+        data = torch.randn(3 * sz, device=device)
+
+        def _test(op, output, left, right):
+            output_exp = torch.empty_like(output)
+            op(left, right, out=output_exp)
+            self.assertEqual(op(left, right, out=output),
+                             output_exp, op.__name__)
+
+        # output is identical to left and right:
+        _test(op, output=data[0:sz], left=data[0:sz], right=data[0:sz])
+        # output, left and right are independent:
+        _test(op, output=data[0:sz],
+              left=data[sz:2 * sz], right=data[2 * sz:3 * sz])
+        # output and left are identical but right is independent:
+        _test(op, output=data[0:sz], left=data[0:sz], right=data[2 * sz:3 * sz])
+        # output and right are identical but left is independent:
+        _test(op, output=data[0:sz], left=data[sz:2 * sz], right=data[0:sz])
+        # left and right have left partial overlap but output is independent:
+        _test(op, output=data[0:sz],
+              left=data[sz:2 * sz], right=data[sz + 1:2 * sz + 1])
+        # output has partial overlap with left:
+        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
+            _test(op, output=data[0:sz],
+                  left=data[1:sz + 1], right=data[2 * sz:3 * sz])
+        # output has partial overlap with right:
+        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
+            _test(op, output=data[0:sz],
+                  left=data[0:sz], right=data[2:sz + 2])
 
     @unittest.expectedFailure
     def test_abs_unary_mem_overlap(self):
-        self.unary_check_mem_overlap(lambda t: t.abs_())
+        self.unary_check_internal_mem_overlap(lambda t: t.abs_())
```

Comment:
Would be good to add these (abs, sinh, cosh) to the table in `test_unary_out_op_mem_overlap`

Reply:
addressed